src/checkers/inference/solver/frontend/VariableCombos.java

```diff
@@ -25,9 +26,13 @@
 public abstract class VariableCombos<T extends Constraint, S> {
 
     private final S emptyValue;
+    private final S contradictoryValue;
+    private final Lattice lattice;
 
-    public VariableCombos(S emptyValue) {
+    public VariableCombos(S emptyValue, S contradictoryValue, Lattice lattice) {
         this.emptyValue = emptyValue;
+        this.contradictoryValue = contradictoryValue;
```

Comment:
Mier, here are my opinion on this `contradictoryValue`, I vote an opposite on this change. Reasons:

1. Efficiency: we can actually early abort/exit if we found a constraint with constant slots is contradict. There is no need to passing a contradictory value into solver to enforce solver unsatisfiable.

2. Reasoning: if we substituted the contradictory constant constraint with an arbitrary contradictory value, then implicitly solver lost the reasoning of this unsatisfiable case ---- the minimized core set reported by solver would be an arbitrary contradictory value. --- even worse, what if this pre-set contradictory value is semantically duplicated with a real contradictory case? (i.e. what if some other cases generated the same core set with this arbitrary contradictory value). Then we cannot distinct if the reasoning represents a contradictory constant constraint, or other cases.

So here I suggest to remove this `contradictoryValue`, and just simply early exit with `Unsatisfiable` if we found a contradictory constant constraint.